siad/daemon_test.go

```diff
@@ -1,24 +1,144 @@
 package main
 
-func testingDaemon() (d *daemon, err error) {
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"testing"
+
+	"github.com/NebulousLabs/Sia/consensus"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/network"
+)
+
+var (
+	APIPort int = 9020
+	RPCPort int = 9120
+)
+
+type daemonTester struct {
+	*daemon
+	*testing.T
+	rpcChan chan struct{}
+}
+
+func newDaemonTester(t *testing.T) *daemonTester {
 	dc := DaemonConfig{
-		APIAddr: ":9020",
-		RPCAddr: ":9021",
+		APIAddr: ":" + strconv.Itoa(APIPort),
+		RPCAddr: ":" + strconv.Itoa(RPCPort),
 
 		HostDir: "hostDir",
 
-		Threads: 2,
+		Threads: 1,
 
 		DownloadDir: "downloadDir",
 
-		WalletDir: "walletDir",
+		WalletDir: "walletDir" + strconv.Itoa(APIPort),
 	}
 
-	d, err = newDaemon(dc)
+	d, err := newDaemon(dc)
 	if err != nil {
-		return
+		t.Fatal("Could not create daemon:", err)
 	}
+	dt := &daemonTester{d, t, make(chan struct{})}
+
+	// overwrite RPCs with special testing RPCs
+	dt.network.RegisterRPC("AddMe", dt.addMe)
+	dt.network.RegisterRPC("RelayBlock", dt.relayBlock)
+	dt.network.RegisterRPC("AcceptTransaction", dt.acceptTransaction)
+
+	go func() {
+		listenErr := d.listen()
+		if listenErr != nil {
+			t.Fatal("API server quit:", listenErr)
+		}
+	}()
+	APIPort++
+	RPCPort++
+
+	return dt
+}
+
+func (dt *daemonTester) address() network.Address {
+	return dt.network.Address()
+}
+
+func (dt *daemonTester) addMe(peer network.Address) error {
+	err := dt.gateway.AddMe(peer)
+	dt.rpcChan <- struct{}{}
+	return err
+}
+
+func (dt *daemonTester) relayBlock(b consensus.Block) error {
+	err := dt.gateway.RelayBlock(b)
+	//dt.rpcChan <- struct{}{}
+	return err
+}
+
+func (dt *daemonTester) acceptTransaction(t consensus.Transaction) error {
+	err := dt.tpool.AcceptTransaction(t)
+	dt.rpcChan <- struct{}{}
+	return err
+}
 
-	go d.listen(dc.APIAddr)
+// mineMoney mines 5 blocks, enough for the coinbase to be accepted by the
+// wallet. This may take a while.
```

Comment:
testing difficulty is set to ~4 hashes per block. Mining 5 blocks is still really fast. The bottleneck might actually be calling AcceptBlock()